cdc/entry/schema_storage.go

```diff
@@ -678,79 +539,118 @@ func (s *Storage) CloneTables() map[uint64]TableName {
 	return mp
 }
 
-// Clone clones Storage
-func (s *Storage) Clone() *Storage {
-	n := &Storage{
-		tableIDToName:  make(map[int64]TableName),
-		tableNameToID:  make(map[TableName]int64),
-		schemaNameToID: make(map[string]int64),
-
-		schemas: make(map[int64]*timodel.DBInfo),
-		tables:  make(map[int64]*TableInfo),
+// SchemaStorage stores the schema information with multi-version
+type SchemaStorage struct {
+	snaps      []*schemaSnapshot
+	snapsMu    sync.RWMutex
+	gcTs       uint64
+	resolvedTs uint64
+}
 
-		truncateTableID:     make(map[int64]struct{}),
-		ineligibleTableID:   make(map[int64]struct{}),
-		version2SchemaTable: make(map[int64]TableName),
-	}
-	for k, v := range s.tableIDToName {
-		n.tableIDToName[k] = v
+// NewSchemaStorage creates a new schema storage
+func NewSchemaStorage(jobs []*timodel.Job) (*SchemaStorage, error) {
+	snap := newEmptySchemaSnapshot()
+	for _, job := range jobs {
+		log.Info("init handle ddl job", zap.Reflect("job", job))
+		if err := snap.handleDDL(job); err != nil {
+			return nil, errors.Trace(err)
+		}
 	}
-	for k, v := range s.tableNameToID {
-		n.tableNameToID[k] = v
+	return &SchemaStorage{
+		snaps:      []*schemaSnapshot{snap},
+		gcTs:       snap.currentTs,
+		resolvedTs: snap.currentTs,
+	}, nil
+}
+
+func (s *SchemaStorage) getSnapshot(ts uint64) (*schemaSnapshot, error) {
+	gcTs := atomic.LoadUint64(&s.gcTs)
+	if ts < gcTs {
+		return nil, errors.Errorf("can not found schema snapshot, the specified ts(%d) is less than gcTS(%d)", ts, gcTs)
 	}
-	for k, v := range s.schemaNameToID {
-		n.schemaNameToID[k] = v
+	resolvedTs := atomic.LoadUint64(&s.resolvedTs)
+	if ts > resolvedTs {
+		return nil, errors.Annotatef(model.ErrUnresolved, "can not found schema snapshot, the specified ts(%d) is more than resolvedTs(%d)", ts, resolvedTs)
 	}
-	for k, v := range s.schemas {
-		n.schemas[k] = v.Clone()
+	s.snapsMu.RLock()
+	defer s.snapsMu.RUnlock()
+	i := sort.Search(len(s.snaps), func(i int) bool {
+		return s.snaps[i].currentTs > ts
+	})
+	if i <= 0 {
+		return nil, errors.Errorf("can not found schema snapshot, ts: %d", ts)
 	}
-	for k, v := range s.tables {
-		n.tables[k] = v.Clone()
+	return s.snaps[i-1], nil
+}
+
+// GetSnapshot returns the snapshot which of ts is specified
+func (s *SchemaStorage) GetSnapshot(ts uint64) (*schemaSnapshot, error) {
+	var snap *schemaSnapshot
+	err := retry.Run(10*time.Millisecond, 25,
+		func() error {
+			var err error
+			snap, err = s.getSnapshot(ts)
+			if errors.Cause(err) != model.ErrUnresolved {
+				return backoff.Permanent(err)
+			}
+			return err
+		})
+	switch err.(type) {
+	case *backoff.PermanentError:
+		return nil, errors.Annotate(err, "timeout")
+	default:
+		return snap, err
 	}
-	for k, v := range s.truncateTableID {
-		n.truncateTableID[k] = v
+}
+
+// GetLastSnapshot returns the last snapshot
+func (s *SchemaStorage) GetLastSnapshot() *schemaSnapshot {
+	s.snapsMu.RLock()
+	defer s.snapsMu.RUnlock()
+	return s.snaps[len(s.snaps)-1]
```

Comment:
We must guarantee the `s.snaps` is not empty 🤔️

Reply:
yes, when we create a SchemaStorage, we will create a snap into `snaps`.
and the `DoGC` function will guarantee that `len(s.sanps)` at least one more